### PR TITLE
Add router and supervisor

### PR DESCRIPTION
**Pull request by Ch0live**
## Description

With this change we enable the ability to pick agents based on a short and concise task. I have made up a list of example agents for the `router.py:pick_agent()` to call, giving it nothing more than the name to make it's decision. It currently will only try to solve the first task and by solve I mean it will pick the agent it thinks is best for the first task. This decision is returned to the user for now on the front end to show some of the decision making.

image

## Changelog
- Moved `create_task_agent.py` prompts into their own Jinja2 template [`create-tasks.j2`](https://github.com/WaitThatShouldntWork/InferGPT/compare/release/sum-my-subscriptions...feature/add-router-and-supervisor?expand=1#diff-43ef87b7355c31381fc92e130a262c95fcfe1108675b71bf061b3b83c775eec0)
- Updated `api.py` to more robustly find the config file using `os.path` package
- Add `router.py` to pick an agent based on task passed
- Add `supervisor.py` to take a list of tasks, take the first and ask `router.py` to solve said first task
- Add `get_response_three_prompts.py` to `llm.py` to allow 3 prompts to be passed when picking an agent
- Set temperature to 0 for all calls to make outcomes more consistent

### CODE REVIEW

File: backend/utils/llm.py [hunk 1]
```diff
@@ -40,6 +40,30 @@ def get_response(system_prompt, user_prompt, tools=None) -> ChatCompletionRespon
...
         tools=tools,
         tool_choice=tool_choice,
+        temperature=0
     )
-    logger.info('{0} response : "{1}"'.format(config.mistral_model, response.choices[0].message.content))
+    logger.debug('{0} response : "{1}"'.format(config.mistral_model, response.choices[0].message.content))
     return response
+
+
+# TODO: Refactor - 1 get_response method and 1 call_model method
```
Comment (harryBedfordSL) [inline backend/utils/llm.py:49]: hmm yeah I wonder whether we need some sort of `message_builder` to create the messages, with how ever many system prompts needed, and then could just pass `messages` into `get_response`. Is your intention to do all these refactors in the wire up stage?

**Review by harryBedfordSL — CHANGES_REQUESTED** [review]
Nice work! Just some minor questions and points from me

File: backend/agents/create_tasks_agent.py [hunk 2]
```diff
@@ -1,41 +1,25 @@
...
-    each task in order to find the best way possible for tasks to answer the overall question.
-
-    If you are unable to breakdown the question into tasks, say that unfortunately you are unable to break down
-    the question.
-
-    You are connected to a neo4j database, as a source of information. You do not need to create a task to connect
-    or retrieve user data from the database.
+def create_tasks(user_prompt: str) -> str:
+    agents = "unresolvable_task, database_agent, fiancial_advisor_agent, web_search_agent" # TODO: Make single source of agent choice knowledge
+    engine = PromptEngine()
```
[the remark this answers is not in the document]
Comment (Ch0live) [reply backend/agents/create_tasks_agent.py:16]: I believe it does, becoming a "module-level variable". Moved outside the method 👍

File: backend/director.py [hunk 3]
```diff
@@ -41,16 +40,18 @@
...
+    logging.debug("Received utterance: {question}")
 
-    logger.info("director calling call_model function")
-    return call_model(director_prompt, user_prompt=question)
+    if determine_intention(question) == "TRUE":
+        tasks_from_question_json = create_tasks(question)
+        final_answer = solve_all_tasks(tasks_from_question_json)
+        return final_answer
 
+    else:
```
[the remark this answers is not in the document]
Comment (Ch0live) [reply backend/director.py:51]: Refactored

File: backend/prompts/prompting.py [hunk 4]
```diff
@@ -4,7 +4,7 @@
 from jinja2 import Environment, FileSystemLoader
 
 
-class PromptEngine:
+class PromptEngine(object):
```
[the remark this answers is not in the document]
Comment (Ch0live) [reply backend/prompts/prompting.py]: It's a leftover change I made trying to get `PromptEngine` to import in the `router.py`. In the end I kept the import sem-relative with a `TODO:` added

```python
from prompts.prompting import PromptEngine # TODO: remove need for .prompting
```

I have now removed it

File: backend/prompts/templates/create-tasks.j2 [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+You are connected to a neo4j database, as a source of information. You do not need to create a task to connect
+or retrieve user data from the database.
+
+you should provide the tasks in valid JSON format following this template:
+
+{
+    "tasks": [
+        {
+            "number": <descending-order-of-the-tasks>,
```
[the remark this answers is not in the document]
Comment (Ch0live) [reply backend/prompts/templates/create-tasks.j2:29]: When first writing this I couldn't remember if JSON kept lists ordered or not. However I am now taking the first task using a 0 index in `supervisor.py` and not even utilizing the numberings.

I have removed this property

File: backend/router.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+import logging
+from utils import get_response_three_prompts
+from prompts.prompting import PromptEngine # TODO: remove need for .prompting
+
+# TODO: Create pick_agent test with mocked calls
+
+def pick_agent(task_string):
+    logging.debug("Picking agent for task: " + task_string)
+
+    prompt_engine = PromptEngine()
```
[the remark this answers is not in the document]
Comment (Ch0live) [reply backend/router.py:11]: Refactored

File: backend/utils/llm.py (hunk 1 above)
Comment (Ch0live) [reply backend/utils/llm.py:49]: Or perhaps even later. I just want to get to the PoC point as soon as possible now, knowing we may scrap parts of it anyway. A bad attitude to writing code but potentially an appropriate one here as we're still not 100% on the feasibility of this approach overall

Comment (Ch0live) [reply backend/utils/llm.py:49]: This also extends to the poor quality testing I've done here, all to be bolstered and improved if we keep this approach and (as a result) these files

File: backend/agents/create_tasks_agent.py [hunk 7]
```diff
@@ -1,41 +1,28 @@
...
-    will provide an answer for the user question.
-
-    To determine the tasks you will create, you will use reasoning, any context you have and will attempt to criticise
-    each task in order to find the best way possible for tasks to answer the overall question.
+def create_tasks(user_prompt: str) -> str:
 
-    If you are unable to breakdown the question into tasks, say that unfortunately you are unable to break down
-    the question.
+    # TODO: Make single source of agent choice knowledge
+    agents = "unresolvable_task, database_agent, fiancial_advisor_agent, web_search_agent"
```
Comment (DannyNicholas) [inline backend/agents/create_tasks_agent.py:15]: This contains an agent called `unresolvable_task`. The `create-tasks.j2` prompt template also appends another agent called `unresolvable_agent`.

Are these intentionally different agents?

File: backend/utils/llm.py (hunk 1 above)
Comment (harryBedfordSL) [reply backend/utils/llm.py:49]: yeah very good points, just me having refactoring ideas, not for now

**Review by harryBedfordSL — APPROVED** [review]
Looks great to me, nice one!

**Review by DannyNicholas — DISMISSED** [review]

File: backend/agents/create_tasks_agent.py (hunk 7 above)
Comment (Ch0live) [reply backend/agents/create_tasks_agent.py]: No 🤦 fixed

**Review by DannyNicholas — APPROVED** [review]

**Review by harryBedfordSL — APPROVED** [review]

**Pull request merged**